### ISSUE TITLE
.NET Core 2.0.6 Method not found: 'System.Span`1<!!1> System.MemoryExtensions.NonPortableCast(System.Span`1<!!0>)'.

### ISSUE THREAD

**Issue by nfdevil** [SixLabors/ImageSharp#510]
### Description
Using an ASP.NET Core 2.0.6 app, I'm getting a thumbnailphoto (bytes array) from our AD and using `Image.Load(thumbnailPhoto)` to load it in a variable. This produces the error:
```
"ClassName": "System.MissingMethodException",
    "Message": "Method not found: 'System.Span`1<!!1> System.MemoryExtensions.NonPortableCast(System.Span`1<!!0>)'.",
    "Data": null,
    "InnerException": null,
    "HelpURL": null,
    "StackTraceString": "   at SixLabors.ImageSharp.Memory.ArrayPoolMemoryManager.Buffer`1.get_Span()\r\n   at SixLabors.ImageSharp.Image.<>c__DisplayClass0_0.<InternalDetectFormat>b__0(IImageFormatDetector x)\r\n   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()\r\n   at System.Linq.Enumerable.TryGetLast[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)\r\n   at SixLabors.ImageSharp.Image.InternalDetectFormat(Stream stream, Configuration config)\r\n   at SixLabors.ImageSharp.Image.DiscoverDecoder(Stream stream, Configuration config, IImageFormat& format)\r\n   at SixLabors.ImageSharp.Image.Decode[TPixel](Stream stream, Configuration config)\r\n   at SixLabors.ImageSharp.Image.WithSeekableStream[T](Configuration config, Stream stream, Func`2 action)\r\n   at SixLabors.ImageSharp.Image.Load[TPixel](Configuration config, Stream stream, IImageFormat& format)\r\n   at SixLabors.ImageSharp.Image.Load[TPixel](Configuration config, Byte[] data)\r\n   at SharePointApi.Controllers.AdController.CreateImageThumbnailWithImageSharp(Byte[] thumbnailPhoto, Int32 width, Int32 height) in F:\\dev\\SharePointApi\\SharePointApi\\Controllers\\AdController.cs:line 174\r\n   at SharePointApi.Controllers.AdController.<GetUsersInAdGroup>d__2.MoveNext() in F:\\dev\\SharePointApi\\SharePointApi\\Controllers\\AdController.cs:line 103",
    "RemoteStackTraceString": null,
    "RemoteStackIndex": 0,
    "ExceptionMethod": null,
    "HResult": -2146233069,
    "Source": "SixLabors.ImageSharp",
    "WatsonBuckets": null,
    "MMClassName": null,
    "MMMemberName": null,
    "MMSignature": null
}
```
It is working however in a .NET Framework 4.5.2 app using the SixLabors.ImageSharp version 1.0.0-beta0002 and SixLabors.Core version 1.0.0-beta0004
### Steps to Reproduce
1. Get a byte array of the thumbnail from AD, saved in the variable: `byte[] thumbnailPhoto`
2. Saving the Image<Rgba32> in a variable with: `using (var image = Image.Load(thumbnailPhoto))`
3. Entering the catch wiht the exception above

### System Configuration

- ImageSharp version: tested both 1.0.0-beta0003 and 1.0.0-dev000956
- Other ImageSharp packages and versions: 
- Environment (Operating system, version and so on): Windows 10 Enterprise 
- .NET Framework version: Microsoft.NetCore.App: 2.0.6 and Microsoft.AspNetCore.All: 2.0.6
- Additional information:

**Comment by iamcarbon** [comment]
Are you able to check that your bin contains the System.Memory 4.5.0-preview1-26216-02 library? 

Does your project contain any references to the old System.Memory library (4.4.0-preview1-25305-02)?

**Comment by nfdevil** [comment]
@iamcarbon here are my dependencies of the project:
![image](https://user-images.githubusercontent.com/7434048/38074589-805368f4-332f-11e8-8c6f-3722af68f2a7.png)

**Comment by JimBobSquarePants** [comment]
@nfdevil Well there's your problem.

The current System.Memory preview version we are depending on is the one on Nuget

[System.Memory 4.5.0-preview1-26216-02](https://www.nuget.org/packages/System.Memory/4.5.0-preview1-26216-02)

The version you have in your solution is from the Myget feed which must have breaking changes.

**Comment by nfdevil** [comment]
Indeed, I've reverted my System.DirectoryServices.AccountManagement to 4.5.0-preview1-26216-02 and reinstalled the ImageSharp nuget. It works now, thanks!

**Comment by JimBobSquarePants** [comment]
Great! I'll close this off now then. Enjoy!

**Comment by xakep139** [comment]
@JimBobSquarePants I've got same exception on dotnet 2.1-preview2. Here's my .csproj file:
```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp2.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="wwwroot\" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.1.0-preview2-final" />
    <PackageReference Include="SixLabors.ImageSharp" Version="1.0.0-beta0003" />
  </ItemGroup>
</Project>
```

**Comment by xakep139** [comment]
In deps.json:
```
"SixLabors.ImageSharp/1.0.0-beta0003": {
  "dependencies": {
    "SixLabors.Core": "1.0.0-beta0005",
    "System.Buffers": "4.5.0-preview2-26406-04",
    "System.Memory": "4.5.0-preview2-26406-04",
    "System.Runtime.CompilerServices.Unsafe": "4.5.0-preview2-26406-04"
  },
  "compile": {
    "lib/netstandard2.0/SixLabors.ImageSharp.dll": {}
  }
}
```

**Comment by JimBobSquarePants** [comment]
@xakep139 beta3 targets the preview1 dependencies. There's breaking changes in preview2.

**Comment by sebastienros** [comment]
What's the plan to move to preview2 ?

**Comment by JimBobSquarePants** [comment]
The nightlies already are

**Comment by sebastienros** [comment]
Well you are too fast for me, I found it and couldn't remove my comment in time :/

**Comment by JimBobSquarePants** [comment]
Haha. You caught me at a good time :smile:

**Comment by HakanL** [comment]
Is there a milestone for when this will be supported via the main NuGet feed? Beta-0004?

**Comment by JimBobSquarePants** [comment]
Next 24 hours hopefully

**Comment by jakenuts** [comment]
Is it expected that this should persist with ASP.NET Core 2.1 RTM & ImageSharp.Web from the standard Nuget feed? The System.Memory is now listed is just 4.5.0.

**Comment by JimBobSquarePants** [comment]
Nuget ImageSharp.Web has not been updated yet and still targets the beta3 of ImageSharp. You probably had the old version of ImageSharp installed

**Comment by knizkar** [comment]
I think `MemoryExtensions.NonPortableCast` has been replaced by `MemoryMashal.Cast` in .NET Core 2.1, see https://github.com/dotnet/coreclr/pull/16091/commits/93ec62347a4e97b920513296eefb27bdacbecfe9.

**Comment by JimBobSquarePants** [comment]
Yup, and if you look at our source you'll see we use `MemoryMashal.Cast` throughout.

**Comment by martinpetrovaj** [comment]
Hey guys, ever since upgrading to the now-released ASP.NET Core 2.1, I'm getting the abovementioned error whenever I try to upload an image to my website. I've seen several issues and a pull request referencing this and I apologize if I'm just too stupid or tired to understand, but what's the status of this issue now that 2.1 is released?

I'm currently writing a step-by-step tutorial on ASP.NET Core and we've chosen ImageSharp as the best way of processing images there currently is, but we needed to upgrade to 2.1 and the version on NuGet doesn't seem to like that ATM. If I understood it correctly, the nightly build on MyGet should already have this covered. Can we expect the fix to arrive on NuGet as well in the near future, or do you think it will take a longer time? (absolute majority of our readers are just hobbyists and juniors capable of only using NuGet) Thanks in advance.

**Comment by JimBobSquarePants** [comment]
You’re using beta4 yeah? We could push out another beta if necessary easily enough

**Comment by antonfirsov** [comment]
@martinpetrovaj how do your project's dependency graph look like? Which versions are actually pulled by nuget and copied to your `bin`?

System.Memory hasn't changed between RC1 and the official 1.0 (the referenced PR is already part of the RC1). Are you sure you don't have other packages pulling in pre-RC1 versions of that package?

Can you make sure you updated all the `System.*` references to the official 1.0 everywhere?

**Comment by martinpetrovaj** [comment]
Sorry, but I can't generate the dependency graph with the VS version I'm using. I've checked the project's and solution's dependencies and packages and everything seems to be using .NET Core 2.1.0 SDK and System.Memory version should be 4.5. Don't know if it helps, all I could find about it in my project's files was this in the deps.json:

```
"System.Memory/4.5.0": {
      "type": "package",
      "serviceable": true,
      "sha512": "sha512-P56/L4B9vhz0YFTC4qIXFgx+VgMhg3JhCm5/EASihTotyjwmBt2QEAQdmd2tNctYET/w0CRJzVH1cwxgbKGnDQ==",
      "path": "system.memory/4.5.0",
      "hashPath": "system.memory.4.5.0.nupkg.sha512"
    }
```

NuGet also says that 4.5 is the only version installed and used. I use but a few 3rd party packages and none of them have dependencies on earlies version of either .NET or System.Memory.

I tried to find the ImageSharp packages files and I think that I'm indeed using beta 4.

I apologize if I haven't answered your question, I have no previous experience troubleshooting this kind of problems, you'll have to talk to me like to a 6yr old kid…

**Comment by antonfirsov** [comment]
@martinpetrovaj I also feel quite stupid when debugging these package hell issues. The whole thing doesn't make sense to me, and I have no idea how to reproduce it.

Are you running on .NET Core 2.1 runtime?

Can you do us a favor and try pulling our newest MyGet packages? There is some (not too high :sigh:) chance that it may make the issue go away. In that case we will release a new hotfix beta ASAP.

**Comment by martinpetrovaj** [comment]
Yes, I'm running on .NET Core 2.1 runtime, the issue sprung up after upgrading to it. All my projects use Microsfot.NETCore.App 2.1.0 SDK, it's a Core MVC web app if it matters.

I have just downloaded and installed the MyGet packages - ImageSharp and ImageSharp.Drawing dev1425, ImageSharp.Web dev000087, I can confirm that everything is working now.

**Comment by antonfirsov** [comment]
@martinpetrovaj great news, thanks!

@JimBobSquarePants @tocsoft we need to push a   new beta through the whole stack.

**Comment by JimBobSquarePants** [comment]
I wish I knew what was causing this. All our current released stack should be targeting the RC.

**Comment by johnnyoshika** [comment]
I seem to be having the same `MissingMethodException: Method not found: 'System.Span`1` error when trying to use ImageSharp.Web. To correct this problem, I'm trying to use the pre-release version from MyGet, but i can't seem to find it. The feed I'm using is https://www.myget.org/F/imagesharp/api/v3/index.json, and while I see ImageSharp there, I don't see ImageSharp.Web.

**Comment by JimBobSquarePants** [comment]
@johnnyoshika see my comment above.

https://github.com/SixLabors/ImageSharp/issues/510#issuecomment-393911300

We haven't pushed the new ImageSharp.Web yet even on MyGet. There's an open PR I need to review tonight if I can.

**Comment by IEvangelist** [comment]
I'm running into this issue too

__Error Message__

> Method not found: 'System.Span`1<!!1> System.MemoryExtensions.NonPortableCast(System.Span`1<!!0>)'.

__`*.csproj`__

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp2.1</TargetFramework>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Centare.Extensions" Version="1.0.16" />
    <PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.0-preview2-final" />
    <PackageReference Include="Microsoft.AspNetCore.SpaServices.Extensions" Version="2.1.0-preview2-final" />
    <PackageReference Include="SixLabors.ImageSharp.Web" Version="1.0.0-beta0003" />
  </ItemGroup>

</Project>

```

__Dependencies__

![image](https://user-images.githubusercontent.com/7679720/41504294-261d61a4-71b0-11e8-91fa-728bf3fff339.png)

What are we supposed to do to fix this, please help?

**Comment by JimBobSquarePants** [comment]
You need this update ImageSharp to beta4 and your system references also.

**Comment by IEvangelist** [comment]
Are you talking about to this one, the one on the top?

![image](https://user-images.githubusercontent.com/7679720/41504470-5b07fcb2-71b5-11e8-9abf-73bab28930b5.png)

I'd prefer not to mess around with adding a myget feed, as I'm already pointing to pre-release. Any other alternatives?

**Comment by JimBobSquarePants** [comment]
ImageSharp beta4 is on Nuget. You should be able to update to that.

**Comment by IEvangelist** [comment]
I'm not seeing it on nuget.org, here is what I see:

![image](https://user-images.githubusercontent.com/7679720/41504485-e971c0a0-71b5-11e8-8f91-bb80fd91a4fc.png)

**Comment by JimBobSquarePants** [comment]
https://www.nuget.org/packages/SixLabors.ImageSharp/1.0.0-beta0004

ImageSharp not ImageSharp.Web

**Comment by IEvangelist** [comment]
Out of curiosity, are there plans to publish a new ImageSharp.Web with the beta4 bits?

**Comment by JimBobSquarePants** [comment]
Got some work to do on it first but I’ll try tonight. But swamped of late.

**Comment by JimBobSquarePants** [comment]
Though we may hold out for the impending beta 5. This week sometime hopefully

**Comment by IEvangelist** [comment]
No worries, thank you so much for all that you do. This is an amazing project that you should be proud of...I know I would be.

**Comment by JimBobSquarePants** [comment]
Thanks mate! I really appreciate that!

**Comment by RevoluPowered** [comment]
Thanks to whoever opened this up! found out someone else published image sharp under a different username. / non pre-release after correcting the dependency to the beta, it's working! :)

**Comment by JimBobSquarePants** [comment]
Oh no, again?! Can you point me to the Nuget package?

**Comment by RevoluPowered** [comment]
I wrongfully installed Veldrid.ImageSharp thinking it was the same one.

It was more human error here, but yeah I need to read descriptions properly

![image](https://user-images.githubusercontent.com/748770/42812216-d2eb311a-89b4-11e8-9613-4861478f1d5c.png)

**Comment by JimBobSquarePants** [comment]
Ah no, that package is great, it allows Veldrid to load gpu textures via ImageSharp. Just needs an update, I should PR.

**Comment by RevoluPowered** [comment]
That looks awesome, I've been looking for something like that for one of my other projects, outside of work! 

I really like image sharp, good work btw, I can't believe these things are not part of the standard lib

**Comment by RevoluPowered** [comment]
:O just found an interesting bug

gonna open up a ticket

**Comment by RevoluPowered** [comment]
I really appreciate that this library worked the first time that I implemented it, zero exceptions and a properly resized image, I didn't even have to care about the aspect ratio fitting, it just handled it for me.

If I could give this library 5* I would.

**Comment by nim** [comment]
Its a cool, very like this

**Comment by labmorales** [comment]
Any updates on this error? In Windows 10 and  Windows Server 2008 R2 it beta0003 works. But for me this error occurs on Windows Server 2016.

I compiled the ImageSharp.Web (directly from master branch) and it worked without any errors on the same server. It's is possible to use the master branch as a nuget reference while the new beta version is not released? 

I also tried to use ImageSharp.Web beta0003 with ImageSharp beta0005 (installed both as dependecies), but it results in a diferent error:
TypeLoadException: Could not load type 'SixLabors.ImageSharp.Processing.Transforms.ResizeOptions' from assembly 'SixLabors.ImageSharp

**Comment by antonfirsov** [comment]
@labmorales the issue being subject of this thread is fixed. By making sure your project is using only the new official versions of System.Memory, there should be no type load errors related to `System.Span<T>`.

Based on your comment, it seems to me that you are facing a compatibility problem related to ImageSharp.Web. I suggest to join our gitter channel, I'm pretty sure someone will be able to help you.

**Comment by labmorales** [comment]
@antonfirsov thanks for the reply. I read the comments again more carefully and installing the ImageSharp beta0004 worked. I was trying before with beta0005.

**Comment by JimBobSquarePants** [comment]
We don't have a released build for beta 5 yet. I'm waiting on some benchmark results.

**Comment by antonfirsov** [comment]
@JimBobSquarepants to avoid further confusion:

The latest NuGet release of IS.Web is beta0003, which is incompatible with the current IS beta0005. This is the root cause of the issue @labmorales is experiencing. Is this correct?

**Comment by johnnyoshika** [comment]
My experience is the same as @labmorales'. The only combination that works for me is:

```
    <PackageReference Include="SixLabors.ImageSharp" Version="1.0.0-beta0004" />
    <PackageReference Include="SixLabors.ImageSharp.Web" Version="1.0.0-beta0003" />
```

**Comment by JimBobSquarePants** [comment]
> We don't have a released build for beta 5 yet. I'm waiting on some benchmark results.

I really don't want to have to repeat myself again.

**Comment by johnnyoshika** [comment]
@JimBobSquarePants My comment wasn't to ask you to repeat yourself. It was to inform the group that my problem is the same as labmorales', and to show anyone who stumbles upon here what csproj settings works for me.
